src/ripple/app/misc/NegativeUNLVote.cpp

```diff
@@ -0,0 +1,354 @@
+    /*
+     * Voting steps:
+     * -- build a reliability score table of validators
+     * -- process the table and find all candidates to disable or to re-enable
+     * -- pick one to disable and one to re-enable if any
+     * -- if found candidates, add ttUNL_MODIFY Tx
+     */
+
+    // Build NodeID set for internal use.
+    // Build NodeID to PublicKey map for lookup before creating ttUNL_MODIFY Tx.
+    hash_set<NodeID> unlNodeIDs;
+    hash_map<NodeID, PublicKey> nidToKeyMap;
+    for (auto const& k : unlKeys)
+    {
+        auto nid = calcNodeID(k);
+        nidToKeyMap.emplace(nid, k);
+        unlNodeIDs.emplace(nid);
+    }
+
+    hash_map<NodeID, unsigned int> scoreTable;
+    if (buildScoreTable(prevLedger, unlNodeIDs, validations, scoreTable))
+    {
+        // build next nUNL
+        auto nUnlKeys = prevLedger->nUnl();
+        auto nUnlToDisable = prevLedger->nUnlToDisable();
+        auto nUnlToReEnable = prevLedger->nUnlToReEnable();
+        if (nUnlToDisable)
+            nUnlKeys.insert(*nUnlToDisable);
+        if (nUnlToReEnable)
+            nUnlKeys.erase(*nUnlToReEnable);
+
+        hash_set<NodeID> nUnlNodeIDs;
+        for (auto const& k : nUnlKeys)
+        {
+            auto nid = calcNodeID(k);
+            nUnlNodeIDs.emplace(nid);
+            if (nidToKeyMap.find(nid) == nidToKeyMap.end())
+            {
+                nidToKeyMap.emplace(nid, k);
+            }
+        }
+
+        auto const seq = prevLedger->info().seq + 1;
+        purgeNewValidators(seq);
+
+        std::vector<NodeID> toDisableCandidates;
+        std::vector<NodeID> toReEnableCandidates;
+        findAllCandidates(
+            unlNodeIDs,
+            nUnlNodeIDs,
+            scoreTable,
+            toDisableCandidates,
+            toReEnableCandidates);
+
+        if (!toDisableCandidates.empty())
+        {
+            auto n =
+                pickOneCandidate(prevLedger->info().hash, toDisableCandidates);
+            assert(nidToKeyMap.find(n) != nidToKeyMap.end());
+            addTx(seq, nidToKeyMap[n], true, initialSet);
+        }
+
+        if (!toReEnableCandidates.empty())
+        {
+            auto n =
+                pickOneCandidate(prevLedger->info().hash, toReEnableCandidates);
+            assert(nidToKeyMap.find(n) != nidToKeyMap.end());
+            addTx(seq, nidToKeyMap[n], false, initialSet);
+        }
+    }
+}
+
+void
+NegativeUNLVote::addTx(
+    LedgerIndex seq,
+    PublicKey const& vp,
+    bool disabling,
+    std::shared_ptr<SHAMap> const& initialSet)
+{
+    STTx nUnlTx(ttUNL_MODIFY, [&](auto& obj) {
+        obj.setFieldU8(sfUNLModifyDisabling, disabling ? 1 : 0);
+        obj.setFieldU32(sfLedgerSequence, seq);
+        obj.setFieldVL(sfUNLModifyValidator, vp.slice());
+    });
+
+    uint256 txID = nUnlTx.getTransactionID();
+    Serializer s;
+    nUnlTx.add(s);
+    auto tItem = std::make_shared<SHAMapItem>(txID, s.peekData());
+    if (!initialSet->addGiveItem(tItem, true, false))
```

Comment:
As of this pr: https://github.com/ripple/rippled/pull/3348, addGiveItem can take advantage of moves. You can either `std::move(tItem)` or just inline the call to `make_shared`. I'd probably just inline the call since I'd rather not leave a moved-from variable around.

Reply:
Thanks, changed.